File: demo/logging_demo/buggy_script.py
#!/usr/bin/env python3

import os
import sys
import json
import requests
import logging
from datetime import datetime
import time

# Configure logging
logging.basicConfig(
    level=logging.DEBUG,
    format='%(asctime)s - %(name)s - %(levelname)s - %(message)s',
    handlers=[
        logging.FileHandler('demo/logging/application.log'),
        logging.StreamHandler(sys.stdout)
    ]
)

logger = logging.getLogger(__name__)

class DataProcessor:
    def __init__(self):
        self.config = None
        self.api_key = None
        self.database_url = None
        
    def load_config(self):
        """Load configuration from file - this will fail"""
        try:
            logger.info("Loading configuration file...")
            # This file doesn't exist - will cause FileNotFoundError
            with open('config/app_config.json', 'r') as f:
                self.config = json.load(f)
            logger.info("Configuration loaded successfully")
        except FileNotFoundError as e:
            logger.error(f"Configuration file not found: {e}")
            raise
        except json.JSONDecodeError as e:
            logger.error(f"Invalid JSON in configuration file: {e}")
            raise
            
    def fetch_api_data(self):
        """Fetch data from API - this will fail with various errors"""
        try:
            logger.info("Fetching data from API...")
            # Using invalid API endpoint
            response = requests.get(
                "https://api.nonexistent-service.com/v1/data",
                headers={"Authorization": f"Bearer {self.api_key}"},
                timeout=5
            )
            response.raise_for_status()
            return response.json()
        except requests.exceptions.ConnectionError as e:
            logger.error(f"Failed to connect to API: {e}")
            raise
        except requests.exceptions.Timeout as e:
            logger.error(f"API request timed out: {e}")
            raise
        except requests.exceptions.HTTPError as e:
            logger.error(f"HTTP error occurred: {e}")
            raise
            
    def process_data(self, data):
        """Process data - this will fail with type errors"""
        try:
            logger.info("Processing data...")
            results = []
            
            # This will fail because data is None
            for item in data['items']:
                # This will cause KeyError
                processed_item = {
                    'id': item['id'],
                    'name': item['name'],
                    'value': item['nonexistent_key'],  # KeyError
                    'calculated': item['price'] * item['quantity']  # May cause TypeError
                }
                results.append(processed_item)
                
            logger.info(f"Processed {len(results)} items successfully")
            return results
            
        except TypeError as e:
            logger.error(f"Type error during data processing: {e}")
            raise
        except KeyError as e:
            logger.error(f"Missing required key in data: {e}")
            raise
            
    def save_to_database(self, data):
        """Save data to database - this will fail"""
        try:
            logger.info("Saving data to database...")
            # Simulate database connection error
            if not self.database_url:
                raise ConnectionError("Database URL not configured")
                
            # Simulate SQL injection vulnerability (bad practice)
            query = f"INSERT INTO users VALUES ('{data['name']}', '{data['email']}')"
            logger.debug(f"Executing query: {query}")
            
            # Simulate database constraint violation
            raise ValueError("Duplicate key violation: user already exists")
            
        except ConnectionError as e:
            logger.error(f"Database connection failed: {e}")
            raise
        except ValueError as e:
            logger.error(f"Database constraint violation: {e}")
            raise
            
    def cleanup_temp_files(self):
        """Cleanup temporary files - this will partially fail"""
        try:
            logger.info("Cleaning up temporary files...")
            temp_files = [
                '/tmp/nonexistent_file.tmp',
                '/tmp/another_missing_file.tmp',
                '/tmp/permission_denied.tmp'
            ]
            
            for file_path in temp_files:
                try:
                    os.remove(file_path)
                    logger.debug(f"Removed {file_path}")
                except FileNotFoundError:
                    logger.warning(f"Temp file not found: {file_path}")
                except PermissionError:
                    logger.error(f"Permission denied removing: {file_path}")
                    
        except Exception as e:
            logger.error(f"Unexpected error during cleanup: {e}")

def main():
    """Main function that orchestrates the buggy operations"""
    logger.info("=== Starting Buggy Application Demo ===")
    
    processor = DataProcessor()
    
    try:
        # This will fail - config file doesn't exist
        processor.load_config()
        
        # This will fail - API doesn't exist
        api_data = processor.fetch_api_data()
        
        # This will fail - data processing errors
        processed_data = processor.process_data(api_data)
        
        # This will fail - database errors
        processor.save_to_database(processed_data[0])
        
    except Exception as e:
        logger.critical(f"Application crashed with error: {e}")
        logger.exception("Full traceback:")
        
    finally:
        # This will generate warnings
        processor.cleanup_temp_files()
        logger.info("=== Application Demo Completed ===")

if __name__ == "__main__":
    main() 

File: demo/logging_demo/buggy_script.js
#!/usr/bin/env node

const fs = require('fs');
const path = require('path');
const https = require('https');

// Create console logger that also writes to file
const logFile = fs.createWriteStream('demo/logging/javascript_errors.log', { flags: 'a' });

function log(level, message) {
    const timestamp = new Date().toISOString();
    const logMessage = `${timestamp} - ${level} - ${message}\n`;
    console.log(logMessage.trim());
    logFile.write(logMessage);
}

class BuggyWebApp {
    constructor() {
        this.config = null;
        this.users = [];
        this.apiEndpoints = [];
    }

    // This will cause reference errors
    loadConfiguration() {
        log('INFO', 'Loading web application configuration...');
        try {
            // Trying to access undefined variable
            const configPath = nonExistentVariable.configPath;
            this.config = require(configPath);
        } catch (error) {
            log('ERROR', `Configuration loading failed: ${error.message}`);
            throw error;
        }
    }

    // This will cause type errors
    processUserData(userData) {
        log('INFO', 'Processing user data...');
        try {
            // Trying to call method on undefined
            const processedUsers = userData.map(user => {
                return {
                    id: user.id,
                    name: user.name.toUpperCase(), // Will fail if name is null/undefined
                    email: user.email.toLowerCase(),
                    age: parseInt(user.age),
                    preferences: user.preferences.split(','), // Will fail if preferences is not string
                    lastLogin: new Date(user.lastLogin),
                    avatar: user.profile.avatar.url // Nested property access that will fail
                };
            });
            return processedUsers;
        } catch (error) {
            log('ERROR', `User data processing failed: ${error.message}`);
            throw error;
        }
    }

    // This will cause async/promise errors
    async fetchExternalData() {
        log('INFO', 'Fetching data from external APIs...');
        try {
            const promises = this.apiEndpoints.map(async (endpoint) => {
                return new Promise((resolve, reject) => {
                    const req = https.get(endpoint.url, (res) => {
                        let data = '';
                        res.on('data', (chunk) => {
                            data += chunk;
                        });
                        res.on('end', () => {
                            try {
                                // This might fail if response is not JSON
                                const jsonData = JSON.parse(data);
                                resolve(jsonData);
                            } catch (parseError) {
                                reject(new Error(`JSON parsing failed for ${endpoint.url}: ${parseError.message}`));
                            }
                        });
                    });
                    
                    req.on('error', (error) => {
                        reject(new Error(`Request failed for ${endpoint.url}: ${error.message}`));
                    });
                    
                    // Set a very short timeout to force timeout errors
                    req.setTimeout(100, () => {
                        req.abort();
                        reject(new Error(`Request timeout for ${endpoint.url}`));
                    });
                });
            });

            const results = await Promise.all(promises);
            return results;
        } catch (error) {
            log('ERROR', `External data fetching failed: ${error.message}`);
            throw error;
        }
    }

    // This will cause array/object manipulation errors
    generateReport(data) {
        log('INFO', 'Generating application report...');
        try {
            const report = {
                totalUsers: data.length,
                averageAge: data.reduce((sum, user) => sum + user.age, 0) / data.length,
                // This will fail if some users don't have the property
                premiumUsers: data.filter(user => user.subscription.isPremium),
                // This will cause issues with null/undefined dates
                recentLogins: data.filter(user => {
                    const daysSinceLogin = (Date.now() - user.lastLogin.getTime()) / (1000 * 60 * 60 * 24);
                    return daysSinceLogin < 7;
                }),
                // This will fail with null reference
                topCountries: data.reduce((acc, user) => {
                    const country = user.address.country;
                    acc[country] = (acc[country] || 0) + 1;
                    return acc;
                }, {})
            };
            return report;
        } catch (error) {
            log('ERROR', `Report generation failed: ${error.message}`);
            throw error;
        }
    }

    // This will cause file system errors
    saveReport(report, filename) {
        log('INFO', `Saving report to ${filename}...`);
        try {
            // Try to write to a directory that doesn't exist
            const invalidPath = path.join('/nonexistent/directory', filename);
            fs.writeFileSync(invalidPath, JSON.stringify(report, null, 2));
            log('INFO', 'Report saved successfully');
        } catch (error) {
            log('ERROR', `Report saving failed: ${error.message}`);
            throw error;
        }
    }
}

// Main execution with intentional errors
async function main() {
    log('INFO', '=== Starting Buggy JavaScript Application Demo ===');
    
    const app = new BuggyWebApp();
    
    // Set up some invalid API endpoints
    app.apiEndpoints = [
        { url: 'https://jsonplaceholder.typicode.com/nonexistent' },
        { url: 'https://api.invalid-domain-12345.com/data' },
        { url: 'https://httpstat.us/500' } // This will return 500 error
    ];

    try {
        // This will fail - undefined variable
        app.loadConfiguration();
    } catch (error) {
        log('CRITICAL', `Configuration error: ${error.stack}`);
    }

    try {
        // This will fail - processing invalid user data
        const badUserData = [
            { id: 1, name: null, email: 'USER@EXAMPLE.COM', age: '25', preferences: null },
            { id: 2, name: 'Jane', email: null, age: 'invalid', preferences: 'music,movies' },
            { id: 3 } // Missing most properties
        ];
        
        const processedUsers = app.processUserData(badUserData);
        
        // This will fail - generating report with bad data
        const report = app.generateReport(processedUsers);
        
        // This will fail - saving to invalid path
        app.saveReport(report, 'user_report.json');
        
    } catch (error) {
        log('CRITICAL', `Application error: ${error.stack}`);
    }

    try {
        // This will fail - network errors
        const externalData = await app.fetchExternalData();
        log('INFO', `Fetched ${externalData.length} external data sources`);
    } catch (error) {
        log('CRITICAL', `External data error: ${error.stack}`);
    }

    // Intentional memory leak simulation
    const memoryLeakArray = [];
    for (let i = 0; i < 100000; i++) {
        memoryLeakArray.push({
            id: i,
            data: new Array(1000).fill('memory leak data'),
            timestamp: new Date(),
            references: memoryLeakArray // Circular reference
        });
    }

    log('WARNING', `Created potential memory leak with ${memoryLeakArray.length} objects`);
    log('INFO', '=== Buggy JavaScript Application Demo Completed ===');
}

// Handle uncaught exceptions
process.on('uncaughtException', (error) => {
    log('FATAL', `Uncaught Exception: ${error.stack}`);
    process.exit(1);
});

process.on('unhandledRejection', (reason, promise) => {
    log('FATAL', `Unhandled Rejection at: ${promise}, reason: ${reason}`);
    process.exit(1);
});

if (require.main === module) {
    main().catch((error) => {
        log('FATAL', `Main function error: ${error.stack}`);
        process.exit(1);
    });
} 